### PR TITLE
7000 wallet pow difficulty increase

### LINKED ISSUE #7000: Wallet doesn't increase the PoW difficulty as it should
It has [been seen](https://vegaprotocol.slack.com/archives/C02KSKW6Y3F/p1669885572659819) that parties can be banned from submitting TXs and is related to the PoW difficulty not being increased in the wallet.

`2022-12-01 08:33:52.404 ERROR 802745 --- [TaskScheduler-1] com.vega.protocol.api.VegaApiClient      : {"error":"party is banned from sending transactions"} `

We need the wallet to submit correct PoW proofs - if it doesn’t the party gets banned for a number of epochs which would significantly impact any LP that is running a bot.

### PR DESCRIPTION
**Pull request by wwestgarth**
closes #7000

### CODE REVIEW

File: wallet/api/pow/pow.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	store []map[string]uint32
+	size  uint64
+
+	lastBlock uint64 // the highest last block we've counted against
+}
+
+// add increments the counter for the number of times pubkey has sent in a transaction with pow against a particular height.
+func (t *txCounter) add(pubKey string, height uint64) error {
+	fmt.Println("adding", pubKey, height)
+	if height < t.lastBlock && height < t.lastBlock-t.size {
```
Comment (ze97286) [inline wallet/api/pow/pow.go:30]: the first condition seems redundant

File: wallet/api/pow/pow.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	// if the network parameter for past blocks has changed we need to tell the counter so it
+	// can tell us if we're using a now historic block
+	counter.resize(uint64(blockData.SpamPowNumberOfPastBlocks))
+	if err := counter.add(pubKey, blockData.Height); err != nil {
+		p.mu.Unlock()
+		return nil, err
+	}
+
+	// how many transactions has this party sent in this block
+	nSent := counter.count(pubKey, blockData.Height)
```
Comment (ze97286) [inline wallet/api/pow/pow.go:127]: why not have add return the count already?

File: CHANGELOG.md [hunk 3]
```diff
@@ -21,6 +21,8 @@
 - [7021](https://github.com/vegaprotocol/vega/issues/7021) - MTM settlement on trading terminated fix.
 - [7102](https://github.com/vegaprotocol/vega/issues/7102) - Ensure the `api-token init -f` wipes the tokens file
 - [7106](https://github.com/vegaprotocol/vega/issues/7106) - Properties of oracle data sent in non-deterministic order
+- [7000](https://github.com/vegaprotocol/vega/issues/7000) - Wallet honours proof of work difficulty increases
+- [7106](https://github.com/vegaprotocol/vega/issues/7106) - Properties of oracle data sent in non-deterministic order
```
Comment (ze97286) [inline CHANGELOG.md:25]: we probably don't need it twice (see line 23) - bad merge?

Comment (wwestgarth) [reply CHANGELOG.md]: Yea something like that, good spot.

**Review by ValentinTrinque — APPROVED** [review]

**Review by jeremyletang — APPROVED** [review]
Would need a @ValentinTrinque review as well, but looks good so far.

**Pull request merged**